File: Sorting/Code/quickSort.py
# implementation of quick sort in python

def swap(a, b, arr):
    if a!=b:
        tmp = arr[a]
        arr[a] = arr[b]
        arr[b] = tmp

def quick_sort(elements, start, end):
    if start < end:
        pi = partition(elements, start, end)
        quick_sort(elements, start, pi-1)
        quick_sort(elements, pi+1, end)

def partition(elements, start, end):
    pivot_index = start
    pivot = elements[pivot_index]

    while start < end:
        while start < len(elements) and elements[start] <= pivot:
            start+=1

        while elements[end] > pivot:
            end-=1

        if start < end:
            swap(start, end, elements)

    swap(pivot_index, end, elements)

    return end


if __name__ == '__main__':
    N = int(input("Enter the number of elements to be sorted:"))
    elements = []
    for i in range(0,N):
        ele = int(input("Enter Elements of Array:"))
        elements.append(ele)
    quick_sort(elements, 0, len(elements)-1)
    print(f'Sorted array: {elements}')

File: Sorting/Code/quickSort.cpp
// { Driver Code Starts
//Initial Template for C

#include <stdio.h>

/* Function to print an array */
void printArray(int arr[], int size)
{
    int i;
    for (i=0; i < size; i++)
        printf("%d ", arr[i]);
    printf("\n");
}

 // } Driver Code Ends
//User function Template for C
void swap(int *nums,int i,int j){
    int temp = nums[i];
    nums[i]=nums[j];
    nums[j]=temp;
}
//Function to sort an array using quick sort algorithm.
void quickSort(int arr[], int start, int end)
{
    // code here
    if (start<end){
        int part = partition(arr,start,end);
        quickSort(arr,start,part-1);
        quickSort(arr,part+1,end);
    }
}
    
int partition (int arr[], int start, int end)
{
   // Your code here
   int pivot = arr[end];
   int i = start-1;
   for(int j=start;j<end;j++){
       if(arr[j]<pivot){
           i++;
           swap(arr,i,j);
       }
   }
   swap(arr,i+1,end);
   return i+1;
}

// { Driver Code Starts.
int main()
{
    int arr[1000],n,T,i;
    scanf("%d",&T);
    while(T--){
    scanf("%d",&n);
    for(i=0;i<n;i++)
      scanf("%d",&arr[i]);
    quickSort(arr, 0, n-1);
    printArray(arr, n);
    }
    return 0;
}  // } Driver Code Ends